### DISCUSSION TITLE
What breaks if you surpass the limits under documentation/Conventions and limits?

### DISCUSSION THREAD

**Discussion by albinaask** [jrouwe/JoltPhysics#1389] [Q&A] [answered]
What are the plausible stability implications of going beyond the stated max speed limit of 500m/s with bigger objects than 10m?

Is this a hard limit beyond which the stability is awful or is it something that just isn't that well tested?
If I find bugs that lies outside these limits, are these something you and the other devs are interested in fixing or something that is out of scope of this project?

**Comment by jrouwe** [comment, marked as answer]
These numbers can probably be pushed a bit before things really break. Especially if you go larger rather than smaller. There are various 'tolerances' in the code which have been tuned for the specified ranges. Some tolerances have been exposed (e.g. `CollideSettingsBase::mCollisionTolerance` or `PhysicsSettings::mPenetrationSlop`), but quite a few haven't been. E.g. search for FLT_EPSILON, 1.0e-6 or 1.0e-12 as they are quite commonly used. Exposing all of these constants in a meaningful way to the user is very difficult as they require expert knowledge to tune them.

I think if you push beyond the limits on the upper bounds, you will find that you can go quite a bit further before things really start breaking down. E.g. a 100m object will probably work fine, it's just not something I really test for.

On the lower bound, you'll run into issues sooner because the tolerances start to become really big for the object. E.g. the default 'penetration slop' of 2cm will be really noticeable if your objects are in the order of a cm. Under the influence of gravity a cm sized object has a very large chance of tunneling through geometry in a single time step.

How big are you thinking about making your objects? And how fast do you want them to go? Feel free to add examples where things go wrong to this discussion. If they seem fixable, I'll look into it, but there are no guarantees.

The main thing I don't want is that people start using the engine with their units set to cm, since that immediately makes all numbers a factor 100 bigger than what the tolerances were designed for.

**Reply by albinaask** [reply]
That seem reasonable! Ideally I'd like to have 300m^3 objects going at least 1km/s. Would this be to throw a wrench into the machinery do you think? I don't really care that much about no-clipping a coin at that speed... I would care about a joint causing phantom forces though. I'm really early into the development cycle, so nothing has gone wrong yet, but physics libs as concept are somewhat known for their instability in my experience...

**Comment by albinaask** [comment]
I think you can safely remove these constraints. Enabling CCD I've gotten objects ranging from 1cm^3 to without a problem colliding at 1km/s, it only stopped working at 2km/s... objects at 1.2km^2 had no problems colliding at 5km/s...

**Reply by jrouwe** [reply]
I'm going to leave the current recommendations in place as it really depends on what you're doing. E.g. if you have a human sized ragdoll move at 5000 m/s then the next representable velocity in a float is 5000.0005 m/s. This means that the relative velocity between two bodies in that ragdoll cannot be more precise than 0.5 mm/s anymore and that can lead to oscillations / constraint violations.

**Reply by albinaask** [reply]
Very interesting point! Seems reasonable
